### DISCUSSION TITLE
Discrepancy in numbers between cobertura (input) and generated report

### DISCUSSION THREAD

**Discussion by bvgheluwe** [danielpalme/ReportGenerator#715] [Q&A]
When we process a cobertura xml file with ReportGenerator, we notice a discrepancy between the general numbers about covered lines, branches, etc.
Example: 
![image](https://github.com/user-attachments/assets/5af164ab-6578-4f91-a9e7-680b832e8b9a)
The numbers 11494 (covered) and 35927 (coverable) are not found in the `coverage` element of the corresponding cobertura file (`lines-covered` and `lines-valid` attributes):
![image](https://github.com/user-attachments/assets/a781d1f3-f968-4a44-905b-c590b0966d60)

How does ReportGenerator calculate these numbers and why do they differ from Cobertura?

**Comment by danielpalme** [comment]
This wiki page explains some of the metrics:
https://github.com/danielpalme/ReportGenerator/wiki/Calculation-details

I don't know how the coverage tool that you have used to generate the cobertura xml, counts the `lines-covered` number.
ReportGenerator counts all lines with a `hits` attribute greater than 0 as covered.

E.g.
```xml
<line number="20" hits="2" branch="false" />
```

Maybe you can count the line elements with `hits` greater 0 in your coverage file. What number do you get?

**Comment by bvgheluwe** [comment]
For our frontend codebase we generate a cobertura file using Karma and Istanbul.
I counted the lines with the following regexes in Notepad++:

* hit count > 0 => `\<line\snumber="\d*"\shits="[1-9][0-9]*"\sbranch` => 9652
* any hit count => `\<line\snumber="\d*"\shits="[0-9]*"\sbranch` => 29934

The numbers I get are exactly the same as those within the xml file itself:

`<coverage lines-valid="29934" lines-covered="9652" line-rate="0.3224" branches-valid="12516" branches-covered="1833" branch-rate="0.1464" timestamp="1734539251822" complexity="0" version="0.1">`

When I run ReportGenerator (v5.4.1) on this file, then it reports the number of coverable lines as 35742 and the number of covered as 11476
![image](https://github.com/user-attachments/assets/789e8471-4d97-4497-aad7-15a5c58ccb77)
Each class has its own file.

**Comment by danielpalme** [comment]
Can you share the Cobertura file? Then I can have a closer look.

You can send it privately via email: reportgenerator@palmmedia.de

**Comment by bvgheluwe** [comment]
Done.
It's a 3844 kb file. I hope the email isn't blocked.

**Reply by danielpalme** [reply]
Thank you. Will have a look as soon as possible.

**Comment by danielpalme** [comment]
I looked at your coverage file.
Your regex only counts the `<line>` elements that exist in in the `<class>` -> `<lines>` elements.

But additional `<line>` elements exist in the `<class>` -> `<methods>` -> `<method>` -> `<lines>` elements.

In total 39807 `<line>` elements exist in the file. 12201 of them with a hit count > 0.

Not all of the `<line>` elements within the `<methods>` elements are unique. Most of  them also appear in the `<class>` -> `<lines>` elements with the same line number.
_ReportGenerator_ primarily uses  the `<class>` -> `<lines>` -> `<line>` elements.  
Elements from `<class>` -> `<methods>` -> `<method>` -> `<lines>` -> `<line>` are only considered if not present in the first list.
In your file 35742 unique `<line>`elements exist. 11476 of them are covered.
Those numbers are also reported by _ReportGenerator_.
